File: OJ/leet1021/leet1021.cpp
/*
leet1021


*/

#include "../utils/LeetCpp.utils.hpp"

using namespace std;

#define ALL(A) A.begin(),A.end()
using VI=vector<int>;
using VVI=vector<VI>;


class Solution {
public:
    string removeOuterParentheses(const string& s) {
        int f = 0;
        string ans; ans.reserve(s.size());
        for (auto&& ch: s)
        {
            if (ch == ')') f--;
            if (f) ans += ch;
            if (ch == '(') f++;
        }
        return ans;
    }
};


int main(int argc, char const *argv[])
{
    Solution sol;   
    cout << sol.removeOuterParentheses("(()())(())") << endl;
    cout << sol.removeOuterParentheses("(()())(())(()(()))") << endl;
    cout << sol.removeOuterParentheses( "()()") << endl;
    return 0;
}

File: OJ/leet1021/leet1021.py
# leet1021
# %%
from typing import List
from pprint import pprint

def init(dims, filled_value = 0):
    if not dims: return filled_value
    return [init(dims[1:], filled_value) for _ in range(dims[0])]


class Solution:
    def removeOuterParentheses(self, s: str) -> str:
        start = 0
        f = 0
        ans = ''
        for i, ch in enumerate(s):
            if ch == ')': f -= 1 
            else: f += 1
            if f == 0: ans += s[start + 1: i]; start = i + 1

        return ans


pprint(Solution().removeOuterParentheses("(()())(())"))
pprint(Solution().removeOuterParentheses("(()())(())(()(()))"))
pprint(Solution().removeOuterParentheses( "()()"))


